pkg/controller/baremetalhost/baremetalhost_controller.go

```diff
@@ -768,10 +788,66 @@ func (r *ReconcileBareMetalHost) actionManageReady(prov provisioner.Provisioner,
 	return r.manageHostPower(prov, info)
 }
 
-func (r *ReconcileBareMetalHost) saveStatus(host *metal3v1alpha1.BareMetalHost) error {
+func (r *ReconcileBareMetalHost) saveHostStatus(host *metal3v1alpha1.BareMetalHost) error {
 	t := metav1.Now()
 	host.Status.LastUpdated = &t
-	return r.client.Status().Update(context.TODO(), host)
+
+	if err := r.saveHostAnnotation(host); err != nil {
+		return err
+	}
+
+	//Refetch host again
+	obj := host.Status.DeepCopy()
+	err := r.client.Get(context.TODO(),
+		client.ObjectKey{
+			Name:      host.Name,
+			Namespace: host.Namespace,
+		},
+		host,
+	)
+	if err != nil {
+		return errors.Wrap(err, "Failed to update Status annotation")
+	}
+	host.Status = *obj
+	err = r.client.Status().Update(context.TODO(), host)
+	return err
+}
+
+func (r *ReconcileBareMetalHost) saveHostAnnotation(host *metal3v1alpha1.BareMetalHost) error {
+	//Repopulate annotation again
+	objStatus := r.getHostStatusFromAnnotation(host)
+	if objStatus == nil {
+		return nil
+	}
```

Comment:
Shouldn't this return an error ? maybe you should modify the getHostStatusFromAnnotation to forward the error upwards.

Reply:
This function is called in two different places and a check of nil is enough IMO. Returning error will unnecessarily complicate the logic and the tests. Cases like no status-annotation initially would throw an error and we should not stop at that error but continue.

Reply:
but what about a malformed annotation? It should block the processing IMO

Reply:
What do mean by malformed annotation? Can you elaborate?

Reply:
Got your point @maelk. I have updated the code accordingly